apps/pollbook/barcode-scanner-daemon/src/unitech_ts100_daemon.rs

```diff
@@ -0,0 +1,607 @@
+            );
+            return Err(err.into());
+        }
+    }
+
+    tokio_serial::new(port_name, baud_rate)
+        .data_bits(DataBits::Eight)
+        .parity(Parity::None)
+        .stop_bits(StopBits::One)
+        .flow_control(FlowControl::None)
+        .timeout(Duration::from_millis(500))
+        .open_native_async()
+        .with_context(|| format!("Failed to open serial port {port_name}"))
+}
+
+async fn write_doc(stream: &mut tokio::net::UnixStream, doc: &AamvaDocument) -> io::Result<()> {
+    let mut buf = Vec::new();
+    serde_json::to_writer(&mut buf, doc)?;
+    buf.push(b'\n');
+
+    stream.write_all(&buf.clone()).await?;
+    stream.flush().await?;
+    Ok(())
+}
+
+/// Writes data to every client in the mutex. Drops any connections that error.
+async fn broadcast_to_clients(
+    clients: &Arc<Mutex<Vec<UnixStream>>>,
+    doc: &AamvaDocument,
+) -> io::Result<()> {
+    let mut guard = clients.lock().await;
+
+    let streams = std::mem::take(&mut *guard);
+    drop(guard);
+
+    let mut alive = Vec::with_capacity(streams.len());
+    for mut stream in streams {
+        match write_doc(&mut stream, doc).await {
+            Ok(()) => alive.push(stream),
+            Err(err) => log!(
+                EventId::SocketClientDisconnected,
+                "Dropping unreachable UDS client due to error: {err}",
+            ),
+        }
+    }
+
+    let mut guard = clients.lock().await;
+    *guard = alive;
+    Ok(())
+}
+
+fn open_socket() -> Result<UnixListener, Error> {
+    // Unlink old socket if it exists
+    let _ = fs::remove_file(UDS_PATH);
+    // Assign address to socket
+    let listener = UnixListener::bind(UDS_PATH)?;
+    fs::set_permissions(UDS_PATH, fs::Permissions::from_mode(0o660))?;
+    log!(
+        event_id: EventId::SocketServerBind,
+        message: format!("UDS bound on {UDS_PATH}"),
+        disposition: Disposition::Success
+    );
+
+    Ok(listener)
+}
+
+/// Reads from any `BufRead`, parses AAMVA documents, and writes JSON+"\n" to any Write.
+/// # Errors
+///
+/// Will return `Err` if an error occurs while reading from USB device.
+pub async fn run_read_write_loop<R>(
+    running: &Arc<AtomicBool>,
+    raw_port: R,
+    clients: Arc<Mutex<Vec<UnixStream>>>,
+) -> Result<(), Error>
+where
+    R: AsyncRead + Unpin + Send + 'static,
+{
+    let max_data_size = MAX_AAMVA_DOCUMENT_SIZE + 1; // + 1 for length of data terminator
+    let mut buf = Vec::with_capacity(max_data_size);
+    let mut start = Instant::now();
+    let mut reader = BufReader::new(raw_port);
+
+    while running.load(Ordering::SeqCst) {
+        if start.elapsed() > HEARTBEAT_LOG_INTERVAL {
+            start = Instant::now();
+            log!(EventId::Heartbeat; EventType::SystemStatus);
+        }
+
+        buf.clear();
+
+        let timeout_result = timeout(
+            READ_LOOP_INTERVAL,
+            reader.read_until(TS100_DATA_TERMINATOR, &mut buf),
+        )
```

Comment:
This timeout isn't strictly necessary because we don't do anything between intervals, but it leaves the door open for device reconnection logic.